Universal/Sniffs/CodeAnalysis/MixedBooleanOperatorSniff.php

```diff
@@ -0,0 +1,107 @@
+<?php
+/**
+ * PHPCSExtra, a collection of sniffs and standards for use with PHP_CodeSniffer.
+ *
+ * @package   PHPCSExtra
+ * @copyright 2021 WoltLab GmbH, 2023 PHPCSExtra Contributors
+ * @license   https://opensource.org/licenses/LGPL-3.0 LGPL3
+ * @link      https://github.com/PHPCSStandards/PHPCSExtra
+ */
+
+namespace PHPCSExtra\Universal\Sniffs\CodeAnalysis;
+
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+/**
+ * Forbid mixing `&&` and `||` within a single expression without making precedence
+ * clear using parentheses.
+ *
+ * @link https://github.com/squizlabs/PHP_CodeSniffer/pull/3205
+ * @link https://github.com/php-fig/per-coding-style/issues/22
+ *
+ * @since 1.2.0
+ */
+final class MixedBooleanOperatorSniff implements Sniff
+{
+
+    /**
+     * Returns an array of tokens this test wants to listen for.
+     *
+     * @since 1.2.0
+     *
+     * @return array<int|string>
+     */
+    public function register()
+    {
+        return [
+            \T_BOOLEAN_OR,
+            \T_BOOLEAN_AND,
+        ];
+    }
+
+    /**
+     * Processes this test, when one of its tokens is encountered.
+     *
+     * @since 1.2.0
+     *
+     * @param \PHP_CodeSniffer\Files\File $phpcsFile The file being scanned.
+     * @param int                         $stackPtr  The position of the current token
+     *                                               in the stack passed in $tokens.
+     *
+     * @return void
+     */
+    public function process(File $phpcsFile, $stackPtr)
+    {
+        $tokens = $phpcsFile->getTokens();
+        $token  = $tokens[$stackPtr];
+
+        $start = $phpcsFile->findStartOfStatement($stackPtr);
+
+        if ($token['code'] === T_BOOLEAN_AND) {
+            $search = T_BOOLEAN_OR;
+        } elseif ($token['code'] === T_BOOLEAN_OR) {
+            $search = T_BOOLEAN_AND;
+        } else {
+            throw new \LogicException('Unreachable');
+        }
+
+        while (true) {
+            $previous = $phpcsFile->findPrevious(
+                [
+                    $search,
+                    T_OPEN_PARENTHESIS,
+                    T_OPEN_SQUARE_BRACKET,
+                    T_CLOSE_PARENTHESIS,
+                    T_CLOSE_SQUARE_BRACKET,
+                ],
+                $stackPtr,
+                $start
+            );
+
+            if ($previous === false) {
+                break;
+            }
+
+            if ($tokens[$previous]['code'] === T_OPEN_PARENTHESIS
+                || $tokens[$previous]['code'] === T_OPEN_SQUARE_BRACKET
+            ) {
+                // We halt if we reach the opening parens / bracket of the boolean operator.
+                return;
+            } elseif ($tokens[$previous]['code'] === T_CLOSE_PARENTHESIS) {
+                // We skip the content of nested parens.
+                $stackPtr = ($tokens[$previous]['parenthesis_opener'] - 1);
```

Comment:
I'm afraid, my terminal doesn't show the underlining when running your command, so I'm unable to see this for myself:

![image](https://github.com/PHPCSStandards/PHPCSExtra/assets/209270/9ca73115-7544-4509-b472-1f94ea8e4041)

Reply:
Would you like me to give you a screenshot of the complete output ?

All the same, the issue with the `$stackPtr` being changed should be clear enough, even without seeing it in the terminal ?

Reply:
> Would you like me to give you a screenshot of the complete output ?

![image](https://github.com/PHPCSStandards/PHPCSExtra/assets/209270/7837b7df-26cc-494f-8f5d-d561d18ace15)

> All the same, the issue with the $stackPtr being changed should be clear enough, even without seeing it in the terminal ?

Yes, it's clearly described. However it would be useful to see the behavior myself when making the adjustments to verify the adjustments are actually correct or even the best behavior. Currently the code scans *backwards* to find offending boolean operators, which in `$foo = true || false || true && false;` probably reports the second `||` which is right in the middle. It might be better to scan forwards, which would reliably report the `&&` which makes more sense.

Reply:
`--colors` fixed it.